Preserve the 'it' variable

phadej opened the pull request.

phadej commented:
ping @sol

sol commented on ghci-wrapper/src/Language/Haskell/GhciWrapper.hs:
```diff
@@ -94,10 +102,12 @@ close repl = do
   hPutStrLn stdin e
+  when preserveIt $ hPutStrLn stdin $ "let " ++ itSafe ++ " = it"
   hPutStrLn stdin marker
+  when preserveIt $ hPutStrLn stdin $ "let it = " ++ itSafe
```
Is this ever harmful or could we just always restore `it`?  If it's ok to always restore `it`, then I vote for removing the `preserveIt` argument from this function and do it unconditionally.

phadej replied:
The `it` variable isn't defined until the first expression is evaluated which sets it, e.g.

```
% ghci
GHCi, version 8.0.2: http://www.haskell.org/ghc/  :? for help
Prelude> it

<interactive>:1:1: error:
    • Variable not in scope: it
    • Perhaps you meant ‘id’ (imported from Prelude)
Prelude> :t 'a'
'a' :: Char
Prelude> it

<interactive>:3:1: error:
    • Variable not in scope: it
    • Perhaps you meant ‘id’ (imported from Prelude)
```

---

Which makes me notice that there's a bug, if the first doctest is e.g. `:t ()`, the `doctest` will fail.

I'll try to fix that shortly.

sol commented:
@phadej Thanks for taking the time to open this PR!  I think this is a useful addition!  Let's try to get this on `master` soon.  Can you please look into the feedback I left.  I also have some comments on the cosmetic side of things.  But let's sort this out first.

phadej commented:
@sol I fixed a bug (`:t 'a'` as a first expression failed), hopefully it illustrates the need for separate `safeEval` and `safeEvalIt` by itself. I'm happy to write a note, if more rationale is needed (not sure where to put it though).

Also I added a new integration test: `it/Setup.hs`, it's a little abuse of the `it`, and again I'm not sure whether
it's specified that `$setup` is run before each test-group and not once per-file.

phadej commented:
Will this be in `doctest-0.11.5`?

sol commented:
On Hackage as `doctest-0.12.0`.

phadej commented:
@sol huge thanks!
